### ISSUE TITLE
file import: Import a sequence of SLR photos with qr codes as markers in the sequence

### ISSUE THREAD

**Issue by urapadmin** [arch-kiosk/arch-kiosk-office#1977]
The idea is to not have the qr codes IN the photo but as a single shot BEFORE a bunch of photos comes that belong to that qr code. 
That way we solve the biggest issue with qr codes entirely: Shooting only the qr code itself in is possible in all kinds of lights and situations, so no more recognition issues.

~The import would work like this:~
- ~The files are ordered chronologically and as soon as the import finds a qr code the information will be applied to all images next in line until another qr code is encountered.´~
- ~When the photographer wants to photograph something else that does not belong to the current group a shot with the lense covered (black image) stops the import from applying the most recent qr code information to the upcoming photos until either another black shot is found or a new qr code is recognized.~

~This might even be compatible with the current import architecture because the photos are processed sequentially. However, I do feel this should be its own import method to make things easier to understand. But it is conceivable to use at least the filename filter as well for this method: one would only have to name the first file of a group of images that show the same context.~

See below, how it is supposed to work. ..

---
- [x] #2138

**Comment by urapadmin** [comment]
- files could also be ordered by the numerical part of their filename. That's what cameras often do
- an option "don't import the qr code file itself" would omit the images that have a qr code
- this is unfortunately also the death of the Raspi project: Who would need it anymore? You could just make a photo of an iPad nearby and then start shooting the object. 😢 
- in additon to the closed camera lense a stop sign (a certain character) that starts a filename could stop a sequence 
- a different character could make a single file in a sequence a candid

**Comment by urapadmin** [comment]
I think this is a bit of a game changer and so I want to try to get it out with USTP and PASU.

**Comment by lbestock** [comment]
I think it is a bit of a game changer, too, and am excited to whip out the cameras once it is testable.

**Comment by urapadmin** [comment]
Okay, that's how we do it:

- a sequence of photos is bracketed by a start qr code and an end qr code, both of which encode the context.
- sequence marker qr codes for sequence import are made in the recording app: #2138 
- files can be either be sorted by the numerical part of their filename (that's what cameras often do) or by the file's creation date.
- The date that will be imported as the date when the image was taken can come from EXIF data otherwise it will be the file's creation date. (The exif date will not be used to order the photos sequentially)
- The images with the qr codes will not be imported themselves but moved out of the way (to the todo folder) if a sequence is complete.
- images outside of a sequence will be ignored (so this kind of import never imports candids or ordinary single shot qr codes).
- qr codes that have a single shot qr code in them that does not belong to the sequence will be ignored.

**Comment by urapadmin** [comment]
### kiosk 1.5.1.0
- the first version of this is installed on the Hillside Demo Server (online servers are not supposed to do it)
- I have not at all tried whether the EXIF date will be used as the time of the shot. Nor have I tested whether anything else is properly used for that datum. So check your date and time of creation field for imported files in the file repository

**Comment by lbestock** [comment]
First test run was successful for importing the sequence files.
But @urapadmin you had intended to remove the bracketing photos themselves into the done folder. They aren't imported, but they are done and shouldn't be run through the next non-sequence import. So, back to you.

**Comment by lbestock** [comment]
Chaos let loose.
If I then import the remainder of the photos using either "bunch" or locally "ton" it assigns the sequence QR codes that are left over in there to the context in the QR code. Not sure if you wanted that to happen even if they hadn't been removed, so I report it.
But the single-shot ones in between sequences imported just fine.

**Comment by urapadmin** [comment]
### Kiosk 1.5.1.1
Addresses:
> you had intended to remove the bracketing photos themselves into the done folder. They aren't imported, but they are done and shouldn't be run through the next non-sequence import. So, back to you.

**Comment by urapadmin** [comment]
> Chaos let loose. If I then import the remainder of the photos using either "bunch" or locally "ton" it assigns the sequence QR codes that are left over in there to the context in the QR code. Not sure if you wanted that to happen even if they hadn't been removed, so I report it. But the single-shot ones in between sequences imported just fine.

What? No, I thought you found out in #2138 that they the sequence qr codes are not imported by the normal ton / upload import. But now they are?

**Comment by urapadmin** [comment]
> > Chaos let loose. If I then import the remainder of the photos using either "bunch" or locally "ton" it assigns the sequence QR codes that are left over in there to the context in the QR code. Not sure if you wanted that to happen even if they hadn't been removed, so I report it. But the single-shot ones in between sequences imported just fine.
> 
> What? No, I thought you found out in #2138 that they the sequence qr codes are not imported by the normal ton / upload import. But now they are?

Ja, that wasn't even implemented. So #2138 must have been a misunderstanding? 

However, the ton and bunch imports should totally skip sequence qr codes. 

This is addressed in 
### Kiosk 1.5.1.2

**Comment by lbestock** [comment]
But that was in 1.5.0, and this was in 1.5.1.

**Comment by urapadmin** [comment]
since I had never implemented it it cannot have worked in 1.5.0, That must have been something that looked right but wasn't. So both things need to be checked.

**Comment by lbestock** [comment]
I didn't have 1.5.1 then. I was testing if it worked properly on the iPad itself, and if Kiosk 1.5.0 refused the new QR code types. And it did, sequence and single-shot.

**Comment by urapadmin** [comment]
### kiosk 1.5.1.3
- change in strategy: Not only should the normal import in the Kiosk versions > 1.5.1 skip the sequence qr codes, it should in fact STOP. 
- please also check what the "bunch" import does if confronted with a sequence qr code (and that it still works on other qr codes)

**Comment by lbestock** [comment]
OK! This works as I expected it to on my next trial run, but I will not say we are done with this yet.

I just re-ran the old photos. It imported all the photos I wanted to (as in, all context photos and the candids, but none of the sequence marking QR codes) in stages that seemed very logical to me:

- First, I ran the sequence. Done, correctly, and then the marking QR codes were in the done folder along with the imported files.
- Then I ran a "ton" import with context unchecked. Got the ones with QR codes into the right contexts, got the ones without the QR codes as candids.

Then I went back and moved a sequence QR code out of the "done" folder back into the main folder and tried a "ton" again, and it was rejected with the error message I expected, one about "hey this is a sequence QR code". 

So, more robust testing needed but the things we had caught so far as not working are now working.
